src/managers/balancer.py

```diff
@@ -80,3 +181,114 @@ def create_internal_topics(self) -> None:
                     ],
                 )
                 logger.info(f"Created topic {topic}")
+
+    def rebalance(self, mode: str, dryrun: bool = True) -> requests.Response:
+        """Triggers a full Kafka cluster partition rebalance."""
+        rebalance_request = self.cruise_control.post(
+            "rebalance", dryrun=dryrun
+        )  # FIXME: will need updating to include add/remove_broker when it works
+        user_task_id = rebalance_request.headers.get("User-Task-ID", "")
+
+        # block entire charm event handling while rebalance in progress
+        while (
+            "Completed" not in self.cruise_control.get_task_status(user_task_id=user_task_id)
+            or self.cruise_control.executing
+        ):
+            logger.info(
+                f"Waiting for '{mode}' rebalance {'dryrun' if dryrun else ''} task to complete..."
+            )
+            time.sleep(10)  # sleep needed as CC API rejects too many requests within a short time
+
+        logger.info(
+            f"'{mode}' rebalance task finished - {self.cruise_control.get_task_status(user_task_id=user_task_id)}"
+        )
+        return rebalance_request
+
+    def _get_storage_size(self, path: str) -> int:
+        """Gets the total storage volume of a mounted filepath, in KB."""
+        return int(self.dependent.workload.exec(f"df --output=size {path} | sed 1d"))
```

Comment:
**note**: It appears that pebble cannot exec commands with pipes. I had to select the second line with python code in the k8s charm

Reply:
Ah right, because the default `kubectl exec` shell is `sh` not `bash`.
You could also solve that with `workload.exec("bash -c 'cat foo.txt | grep bar'")`, that should work.

Reply:
Example from ZK:
```python
        bash_netcat = (
            f"echo '4lw' | (exec 3<>/dev/tcp/localhost/{CLIENT_PORT}; cat >&3; cat <&3; exec 3<&-)"
        )
        ruok = [bash_netcat.replace("4lw", "ruok")]
        srvr = [bash_netcat.replace("4lw", "srvr")]

        # timeout needed as it can sometimes hang forever if there's a problem
        # for example when the endpoint is unreachable
        timeout = ["timeout", "10s", "bash", "-c"]

        try:
            ruok_response = self.exec(command=timeout + ruok)
```

Reply:
Would it help if I added that to this PR?

Reply:
I'll take care of that in the k8s charm first, and revisit this charm